### PR TITLE
ssh: support linking

### PR DESCRIPTION
**Pull request by mroutis**
### Description
Support links for remote SSH cache - close https://github.com/iterative/dvc/issues/1652

### Setup

**SSH**:
```bash
# sudo systemctl start sshd
remote=/home/mroutis/tmp/ssh
rm -rf ${remote} && mkdir ${remote}
cache=$(mktemp --directory --tmpdir=${remote})
dvc init --no-scm --force
dvc remote add ssh ssh://localhost/${cache}
dvc remote modify ssh type symlink
dvc config cache.ssh ssh
echo "foo" > ${remote}/foo
dvc add ssh://localhost/${remote}/foo
```

**Test local links**:
```bash
echo "bar" > bar
dvc add bar
```

### Notes

Linking happens when DVC moves a file from cache to the output path:
  1. `stage.commit`
  2. for out in self.outs: `out.commit`
  3. `if out.use_cache: out.cache.save`
  4. `_save_file`
  5. `remote.base.link(cache_info, path_info)` ← `ipdb.set_trace`
    i) `_link`
    ii) `_try_links` → `slow_link_guard`
    iii) `_get_link_method`
    iv) `_do_link(from_info, to_info, link_method)`

Also, you'll need to implement `_needs_checkout` for `RemoteSSH`
after https://github.com/iterative/dvc/pull/2358 is closed.

Who uses `self.link`?
  * `_checkout_dir`
  * `_checkout_file`
  * `_save_file`

* There's a `link` in remote base that's equivalent to `self.copy`
  and only `RemoteLOCAL` overwrites it.

### TODO
  * ~~[ ] Abstract link logic to `utils.fs` module~~
  * ~~[ ] Delegate `RemoteLOCAL.link` and `RemoteSSH.link` to `fs.link`~~
    * Can't abstract to `utils.fs` module because link is a remote
    specific operation.

  * [x] Diff `remote/base.py` with `remote/local/__init__.py` to
  detect methods related with link operations:
    * Code related with `cache.protected`
    * FS operations based on the remote:
      * `exists`
      * `makedirs`
      * `link`
      * `is_empty`
      * `isfile`
      * `isdir`
      * `getsize`
      * `walk`
      * `remove`
      * `move`
      * `copy`
      * `symlink`
      * `hardlink`
      * `reflink`
  * [x] Adapat `link` logic for `RemoteSSH`
    * `CACHE_TYPE_MAP` operations require an SSH connection.
  * [x] Support the following _link types_ for `RemoteSSH`:
    * [x] `symlink`
    * [x] `reflink`
    * [x] `hardlink`
  * [x] Abstract link logic to `RemoteBASE`
    * [x] Implement the following in every remote (or fallback to copy):
      * `isfile`
      * `makedirs`
      * `getsize`
      * `open`
  * [x] Test manually `RemoteLOCAL` and `RemoteSSH`
  * [x] Write unit tests for `RemoteSSH` and `SSHConnection`
    * [X] Test empty hardlink on SSH
    * [x] Test link methods
  * [x] Write functional test for `symlink` as `mockssh` doesn't support it
  * [ ] Update docs accordingly

### Related
  * [ ] Support protected cache for `RemoteSSH`

  * [ ] Abstract fs operations from `RemoteLOCAL` to a
  `remote.local.fs` or `fs.local` module and use it across all the codebase
  (this is useful because we are supporting different OS and dealing
  with different python versions).

  * [x] Switch to a client/pool for `SSH` instead of passing a
  connection in every operation

  * [ ] `RemoteSSH` already has an `SSHConnection` class that
  abstracts operations related to the remote "file system".
  It would be cool if we concentrate most logic on a `RemoteBASE`
  class and delegate operations to the remote's `fs` module.

  * [x] `link` relies on `path_info` instead of `path`, this doesn't
  match with implementation of `RemoteSSH` because it relies on
  `url_info`.
    * Now `RemoteSSH` uses `path_info`

### Questions
  * [x] Why does `fspath` doesn't work for `URLInfo` ?
    * As @efiop noticed `fspath` is for loading the file system representation
    of a path object.
    * `URLInfo` doesn't represent a file system path
    * You need to pass the `*Info` objects to the remote methods because it's
    how they perform the connection.
    Take Check `remote.ssh.move` as an example:
    ```python
    def move(self, from_info, to_info):
      with self.ssh(from_info) as ssh:
        ssh.move(from_info.path, to_info.path)
    ```
    * Even if the remote has already enough information from the config
    (host, user, port, password, etc.), it extracts the connection info
    from the `URLInfo` object.

  * [x] Why are we relying on `URLInfo` to make the connection and not
  reading from the config file?

  * [x] Do we need a separate config section for `cache.protected`
  and `cache.types` specificaly for `RemoteSSH` or should we use
  the same options for both (`RemoteLOCAL` and `RemoteSSH`)?
    * @efiop: `type` and `protected` are already part of the RemoteSCHEMA

  * [x] Do we need to implement `slow_link_detection` for `RemoteSSH`?
    * Yes

  * [x] Besides `RemoteLOCAL`, all the other remotes uses `posixpath`s, right?
    * Yes

### CODE REVIEW

**Comment by mroutis** [comment]
@efiop , still a WIP but there are some inquiries about the implementation that we can start discussing :slightly_smiling_face:

File: dvc/remote/ssh/__init__.py [hunk 1]
```diff
@@ -47,9 +50,21 @@ class RemoteSSH(RemoteBASE):
     # We use conservative setting of 4 instead to not exhaust max sessions.
     CHECKSUM_JOBS = 4
 
+    # XXX: Friendly reminder to add "reflink" to the default types
+    # after is implemented
+    DEFAULT_CACHE_TYPES = ["symlink", "copy"]
```
Comment (efiop) [inline dvc/remote/ssh/__init__.py]: Actually, how about we leave "copy" by default to provide the same safe behaviour in both local and ssh?

File: dvc/remote/ssh/__init__.py [hunk 2]
```diff
@@ -47,9 +50,21 @@ class RemoteSSH(RemoteBASE):
...
+
     def __init__(self, repo, config):
         super(RemoteSSH, self).__init__(repo, config)
 
+        types = config.get(Config.SECTION_CACHE_TYPE, None)
```
Comment (efiop) [inline dvc/remote/ssh/__init__.py]: Let's define DEFAULT_CACHE_TYPES = ["copy"] in RemoteBASE, since it is indeed what we use by default. And then move this `types` parsing to RemoteBASE as well, so we can avoid duplication in RemoteLOCAL and RemoteSSH.

File: dvc/remote/ssh/__init__.py [hunk 3]
```diff
@@ -278,3 +293,75 @@ def exists_with_progress(chunks):
...
+
+    def _do_link(self, from_info, to_info, link_method, ssh):
+        # XXX: We are testing if file exists rather than if file is a link
+        if ssh.exists(to_info.path):
+            raise DvcException("Link '{}' already exists!".format(to_info))
+        else:
+            link_method(from_info.path, to_info.path)
+
+        if self.protected:
+            self.protect(to_info)
```
Comment (efiop) [inline dvc/remote/ssh/__init__.py]: I'm pretty sure protect is not implemented or does nothing for SSH right now. Does sftp have chmod?

**Comment by efiop** [comment]
> Why does fspath doesn't work for URLInfo ?

URLInfo is about urls, which are not part of the filesystem. fspath is inspored by fspath from cpython https://github.com/python/cpython/blob/master/Lib/os.py 

> Do we need a separate config section for cache.protected
and cache.types specificaly for RemoteSSH or should we use
the same options for both (RemoteLOCAL and RemoteSSH)?

It is a part of REMOTE_SCHEMA already, so you could already use it for any remote section.

> Do we need to implement slow_link_detection for RemoteSSH?

It would be nice, but we could leave it as a separate ticket. Looks like we would be able to simply reuse the already existing guard as is.

>  Besides RemoteLOCAL, all the other remotes uses posixpaths, right?

Yes, but we try to use PathInfo that handles that for us.

File: dvc/remote/ssh/__init__.py [hunk 4]
```diff
@@ -47,9 +50,21 @@ class RemoteSSH(RemoteBASE):
     # We use conservative setting of 4 instead to not exhaust max sessions.
     CHECKSUM_JOBS = 4
 
+    # XXX: Friendly reminder to add "reflink" to the default types
```
Comment (shcheklein) [inline dvc/remote/ssh/__init__.py]: let's not use XXX and TODO in code. Let's create a ticket if something needs to be updated at some point in the future.

Comment (mroutis) [reply dvc/remote/ssh/__init__.py]: this is a reminder for myself, @shcheklein , I'll make sure to remove them when implemented :slightly_smiling_face:

File: dvc/remote/ssh/__init__.py (hunk 3 above)
Comment (Suor) [reply dvc/remote/ssh/__init__.py]: It does and is supported by paramiko - http://docs.paramiko.org/en/2.6/api/sftp.html#paramiko.sftp_client.SFTPClient.chmod

File: dvc/remote/ssh/__init__.py [hunk 5]
```diff
@@ -278,3 +293,75 @@ def exists_with_progress(chunks):
...
+                self._do_link(from_info, to_info, link_method, ssh)
+                return
+
+            except DvcException as exc:
+                msg = "Cache type '{}' is not supported: {}"
+                logger.debug(msg.format(link_types[0], str(exc)))
+                del link_types[0]
+                i -= 1
+
+        raise DvcException("no possible cache types left to try out.")
```
[the remark this answers is not in the document]
Comment (Suor) [reply dvc/remote/ssh/__init__.py]: The other issue is we pass `ssh` here everywhere. This could be solved by switching to transparent client/pool, e.g. an object, which will summon a client from a pool for each method call.

File: dvc/remote/ssh/__init__.py (hunk 3 above)
Comment (mroutis) [reply dvc/remote/ssh/__init__.py]: It does support `chmod`, @efiop

Comment (mroutis) [reply dvc/remote/ssh/__init__.py]: It does support `chmod`, @efiop

File: dvc/remote/ssh/connection.py [hunk 6]
```diff
@@ -282,3 +288,17 @@ def open_max_sftp_channels(self):
                         raise
                     break
         return self._sftp_channels
+
+    def open(self, *args, **kwargs):
+        return self.sftp.open(*args, **kwargs)
+
+    def symlink(self, *args, **kwargs):
+        return self.sftp.symlink(*args, **kwargs)
```
Comment (efiop) [inline dvc/remote/ssh/connection.py]: does sftp.symlink create dirs automatically? I see that you are using self.makedirs in reflink and hardlink, but not here.

Comment (mroutis) [reply dvc/remote/ssh/connection.py]: good observation, I don't think it does, I'll use `makedirs` too.

Comment (efiop) [reply dvc/remote/ssh/connection.py]: @mroutis but isn't link() already running makedirs?

File: dvc/remote/base.py [hunk 7]
```diff
@@ -352,7 +361,89 @@ def changed(self, path_info, checksum_info):
...
+        # (i.e. 68b329da9893e34099c7d8ad5cb9c940), therfore, they will be
+        # linked to the same file in the cache.
+        #
+        # From https://en.wikipedia.org/wiki/Hard_link
+        #   * ext4 limits the number of hard links on a file to 65,000
+        #   * Windows with NTFS has a limit of 1024 hard links on a file
+        #
+        # That's why we simply create an empty file rather than a link.
+        if self.getsize(from_info) == 0:
+            self.open(to_info, "w+").close()
```
Comment (Suor) [inline dvc/remote/base.py]: `w+` is the same as `w` I believe.

File: dvc/remote/base.py [hunk 8]
```diff
@@ -352,7 +361,89 @@ def changed(self, path_info, checksum_info):
...
+        self._try_links(from_info, to_info, link_types)
+
+    @slow_link_guard
+    def _try_links(self, from_info, to_info, link_types):
+        link_methods = {
+            "copy": self.copy,
+            "symlink": self.symlink,
+            "hardlink": self.hardlink,
+            "reflink": self.reflink,
+        }
```
Comment (Suor) [inline dvc/remote/base.py]: Maybe use `getattr()` instead?

File: dvc/remote/ssh/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mroutis) [reply dvc/remote/ssh/__init__.py]: @Suor , any example on how to use the _client/pool_ ? I prefer to use this instead of passing the connection everywhere

File: dvc/remote/ssh/connection.py (hunk 6 above)
Comment (mroutis) [reply dvc/remote/ssh/connection.py]: yes, @efiop , this is something that we should decide on.
If we are going to `makedirs` on `copy`/`hardlink`/`reflink`/`symlink` or we are going to trust the programmer to make sure the destination already has their parent directories created.

This trust thingy is something that happens across the code base, for example, some methods have assertions like the `isfile`, `exists`, and things like that.

It would be interesting to hear other's opinions on this aspect ^ (cc: @Suor , @pared , @shcheklein)

Comment (efiop) [reply dvc/remote/ssh/connection.py]: @mroutis we are already creating dirs automatically on creating a link in link() method where it is justified, I'm just saying that these low-level methods shouldn't do that.

File: dvc/remote/local/__init__.py [hunk 9]
```diff
@@ -136,29 +122,44 @@ def link(self, from_info, to_info):
         self._link(from_info, to_info, self.cache_types)
 
     def _link(self, from_info, to_info, link_types):
```
Comment (efiop) [inline dvc/remote/local/__init__.py]: I guess you meant to remove this one, as you have it in RemoteBASE already, right?

File: dvc/remote/ssh/__init__.py [hunk 10]
```diff
@@ -47,9 +51,19 @@ class RemoteSSH(RemoteBASE):
     # We use conservative setting of 4 instead to not exhaust max sessions.
     CHECKSUM_JOBS = 4
 
+    DEFAULT_CACHE_TYPES = ["reflink, copy"]
+
     def __init__(self, repo, config):
         super(RemoteSSH, self).__init__(repo, config)
 
+        types = config.get(Config.SECTION_CACHE_TYPE, None)
```
Comment (efiop) [inline dvc/remote/ssh/__init__.py]: Seems like you've forgot to remove this too after generalizing it in RemoteBASE :slightly_smiling_face:

File: dvc/remote/ssh/__init__.py [hunk 11]
```diff
@@ -278,3 +325,81 @@ def exists_with_progress(chunks):
                 ret = list(itertools.compress(checksums, in_remote))
                 pbar.update_desc("", 0)  # clear path name description
                 return ret
+
+    def link(self, from_info, to_info):
```
Comment (efiop) [inline dvc/remote/ssh/__init__.py]: And this looks like.

File: dvc/remote/ssh/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Suor) [reply dvc/remote/ssh/__init__.py]: @mroutis if you'll implement all those methods you won't need that I think. Just don't pass `ssh`, summon a connection every time you need.

File: dvc/remote/ssh/__init__.py [hunk 12]
```diff
@@ -162,13 +176,46 @@ def isdir(self, path_info):
...
 
+    def symlink(self, from_info, to_info):
+        if from_info.scheme != self.scheme or to_info.scheme != self.scheme:
+            raise NotImplementedError
+
+        with self.ssh(from_info) as ssh:
+            ssh.symlink(from_info.path, to_info.path)
+
+    def hardlink(self, from_info, to_info):
+        if from_info.scheme != self.scheme or to_info.scheme != self.scheme:
```
Comment (Suor) [inline dvc/remote/ssh/__init__.py]:
```suggestion
        if not from_info.scheme == to_info.scheme == self.scheme:
```

**Comment by mroutis** [comment]
@efiop , updated the description with a new question:

> Why are we relying on URLInfo to make the connection and not
reading from the config file?

File: dvc/remote/base.py [hunk 13]
```diff
@@ -352,7 +363,49 @@ def changed(self, path_info, checksum_info):
...
+                del link_types[0]
+                i -= 1
+
+        raise DvcException("no possible cache types left to try out.")
+
+    def _do_link(self, from_info, to_info, link_method):
+        # XXX: We are testing if file exists rather than if file is a link
+        if self.exists(to_info):
+            raise DvcException("Link '{}' already exists!".format(to_info))
+        else:
```
Comment (efiop) [inline dvc/remote/base.py]: This `else` is not needed.

File: dvc/remote/base.py [hunk 14]
```diff
@@ -352,7 +363,49 @@ def changed(self, path_info, checksum_info):
...
+            except DvcException as exc:
+                msg = "Cache type '{}' is not supported: {}"
+                logger.debug(msg.format(link_types[0], str(exc)))
+                del link_types[0]
+                i -= 1
+
+        raise DvcException("no possible cache types left to try out.")
+
+    def _do_link(self, from_info, to_info, link_method):
+        # XXX: We are testing if file exists rather than if file is a link
```
Comment (efiop) [inline dvc/remote/base.py]: Why would we care if that is a link? This check is needed to ensure that we won't be trying to create a link path for which already exists. You can see us using lexists in RemoteLOCAL. But in RemoteSSH, we are using `stat` instead of `lstat` for `exists`, which should be fixed.

File: dvc/remote/base.py [hunk 15]
```diff
@@ -719,6 +790,7 @@ def _checkout_dir(
             if self.changed(entry_info, entry_checksum_info):
                 if self.exists(entry_info):
                     self.safe_remove(entry_info, force=force)
+                self.makedirs(entry_info.parent)
```
Comment (efiop) [inline dvc/remote/base.py]: Didn't we discuss that this one is not needed, because `link` will call it anyway?

File: tests/unit/remote/ssh/test_connection.py [hunk 16]
```diff
@@ -64,3 +76,22 @@ def test_walk(tmp_path, ssh):
             paths.add(posixpath.join(root, entry))
 
     assert paths == expected
+
+@pytest.mark.skipif(platform.system() != "Darwin",
+                    reason="Only works on OSX (because of APFS)")
+def test_reflink(chdir_tmp, foo, ssh):
+    ssh.reflink("foo", "bar")
+    assert System.is_reflink(bar_path)
```
Comment (efiop) [inline tests/unit/remote/ssh/test_connection.py]: There is no such function. We've discussed that there is no good way to distinguish copy and reflink, right? So just check that bar_path exists and is not a symlink or a hardlink, it will more than enough.

File: tests/unit/remote/ssh/test_connection.py [hunk 17]
```diff
@@ -64,3 +76,22 @@ def test_walk(tmp_path, ssh):
...
+
+@pytest.mark.skip(reason="MockSSH doesn't support symlinks")
```
Comment (efiop) [inline tests/unit/remote/ssh/test_connection.py]: You can at least mock sftp.symlink and check that it is called.

Comment (efiop) [reply tests/unit/remote/ssh/test_connection.py]: Also, you might be able to func test it same as we do in TestRemoteSSH in tests/func/test_data_sync.py.

File: dvc/remote/base.py (hunk 14 above)
Comment (mroutis) [reply dvc/remote/base.py]: Agree with `lstat` :+1:

File: dvc/remote/base.py (hunk 15 above)
Comment (mroutis) [reply dvc/remote/base.py]: Yes! Good catch, @efiop , it was a leftover

File: tests/unit/remote/ssh/test_connection.py (hunk 16 above)
Comment (mroutis) [reply tests/unit/remote/ssh/test_connection.py]: Right :+1:

File: tests/unit/remote/ssh/test_connection.py (hunk 17 above)
Comment (mroutis) [reply tests/unit/remote/ssh/test_connection.py]: ready :+1:

File: dvc/remote/base.py (hunk 14 above)
Comment (mroutis) [reply dvc/remote/base.py]: The current issue with this one is that `mock-ssh-server` doesn't support `lstat` operation and tests are failing. Need to figure out a way to test this

Comment (efiop) [reply dvc/remote/base.py]: @mroutis It is actually pretty simple to implement, take a look at https://github.com/carletes/mock-ssh-server/blob/master/mockssh/sftp.py#L83 . You can fork it into iterative, modify what is needed and submit a patch. And while the patch is being reviewed, we can use our fork.

File: dvc/remote/base.py (hunk 13 above)
Comment (mroutis) [reply dvc/remote/base.py]: Oops, crossed comments:

https://github.com/iterative/dvc/pull/2426#discussion_r319731653

> I was pointing out that we weren't testing for an actual "link" because the error message says something like: `Link 'foo.txt' already exists!` I guess it is `File` instead of `Link` :sweat_smile:

File: dvc/remote/base.py (hunk 14 above)
Comment (mroutis) [reply dvc/remote/base.py]: It's a shame I didn't check this before, @efiop , I ended up using a forked version, indeed :see_no_evil:

File: dvc/remote/base.py (hunk 13 above)
Comment (mroutis) [reply dvc/remote/base.py]: wait, my bad, I got confused; ignore the above one, @efiop

File: dvc/remote/base.py (hunk 14 above)
Comment (mroutis) [reply dvc/remote/base.py]: https://github.com/carletes/mock-ssh-server/pull/9

File: dvc/remote/local/__init__.py [hunk 18]
```diff
@@ -251,6 +171,41 @@ def move(self, from_info, to_info):
 
         move(from_info, to_info, mode=mode)
 
+    def copy(self, from_info, to_info):
+        shutil.copyfile(from_info.fspath, to_info.fspath)
+
+    def symlink(self, from_info, to_info):
+        System.symlink(from_info.fspath, to_info.fspath)
```
Comment (Suor) [inline dvc/remote/local/__init__.py]: `System.symlink()` works with Path-like objects, so pass them as is. The overall approach is:
- our utilities accept Path-like
- `os` utilities only accept them in Python 3.6+, so we wrap into `fspath_py35()`
- if we want to convert to string we use stringification

So `.fspath` should never be used outside `PathInfo` implementation. We don't always follow these rules in current code, but new code should follow.

File: dvc/remote/hdfs.py [hunk 19]
```diff
@@ -143,8 +143,6 @@ def _download(self, from_info, to_file, **_kwargs):
 
     @contextmanager
     def open(self, path_info, mode="r", encoding=None):
-        assert mode in {"r", "rt", "rb"}
-
```
Comment (Suor) [inline dvc/remote/hdfs.py]: Why did you remove this `assert`? It is still working only for those modes. As I see it this PR should only touch local and ssh.

File: dvc/remote/local/__init__.py [hunk 20]
```diff
@@ -251,6 +171,41 @@ def move(self, from_info, to_info):
...
+            self.open(to_info, "w").close()
+
+            logger.debug(
+                "Created empty file: {src} -> {dest}".format(
+                    src=str(from_info), dest=str(to_info)
+                )
+            )
+            return
+
+        System.hardlink(from_info.fspath, to_info.fspath)
```
Comment (Suor) [inline dvc/remote/local/__init__.py]: `System.hardlink()` also works with Path-like objects.

File: dvc/remote/local/__init__.py [hunk 21]
```diff
@@ -251,6 +171,41 @@ def move(self, from_info, to_info):
...
+
+    def reflink(self, from_info, to_info):
+        System.reflink(from_info.fspath, to_info.fspath)
```
Comment (Suor) [inline dvc/remote/local/__init__.py]: Same as above

File: dvc/remote/ssh/__init__.py [hunk 22]
```diff
@@ -207,8 +250,6 @@ def _upload(self, from_file, to_info, name=None, no_progress_bar=False):
 
     @contextmanager
     def open(self, path_info, mode="r", encoding=None):
-        assert mode in {"r", "rt", "rb"}
```
Comment (Suor) [inline dvc/remote/ssh/__init__.py]: You removed assert, but you didn't implement new modes. BTW, it would be easier for you to implement `'wb'` mode then `'w'`, since paramiko files are always binary.

File: dvc/repo/__init__.py [hunk 23]
```diff
@@ -440,6 +440,8 @@ def is_dvc_internal(self, path):
     @contextmanager
     def open(self, path, remote=None, mode="r", encoding=None):
         """Opens a specified resource as a file descriptor"""
+        assert mode in {"r", "rt", "rb"}
```
Comment (Suor) [inline dvc/repo/__init__.py]: This assert guards incomplete implementation, this is why it was placed next to it. Moving it away will cause multiple issues:
- people might think that `SomeRemote.open(..., mode="w")` works
- when adding new remote it will be non-obvious how much functionality to `.open()` one must add
- extending implementation for one remote will lead to incoherence between guard and this new implementation.

File: dvc/utils/http.py [hunk 24]
```diff
@@ -13,8 +13,6 @@ def open_url(url, mode="r", encoding=None):
     Resumes on connection error.
     Url could be a string or a callable returning a string.
     """
-    assert mode in {"r", "rt", "rb"}
```
Comment (Suor) [inline dvc/utils/http.py]: Same as above.

File: tests/func/test_data_cloud.py [hunk 25]
```diff
@@ -449,6 +452,25 @@ def _get_url(self):
...
+
+        remote = self.cloud.get_remote()
+
+        os.remove("bar")
+
+        def urlinfo(path):
+            cwd = os.getcwd()
+            path = posixpath.abspath(os.path.join(cwd, path))
+            netloc = remote.path_info.netloc
+            return URLInfo("ssh://{}{}".format(netloc, path))
```
Comment (Suor) [inline tests/func/test_data_cloud.py]: Should not be needed. Won't this work:
```python
from_info = remote.path_info / "foo"
to_info = remote.path_info / "bar"
```

File: tests/unit/remote/ssh/test_ssh.py [hunk 26]
```diff
@@ -182,3 +185,24 @@ def test_ssh_gss_auth(mock_file, mock_exists, config, expected_gss_auth):
...
+        "keyfile": ssh_server.test_creds["key_filename"],
+    }
+    remote = RemoteSSH(None, config)
+
+    open("empty", "w").close()
+
+    def urlpath(path):
+        netloc = remote.path_info.netloc
+        path = posixpath.abspath(path)
+        return URLInfo("ssh://{}/{}".format(netloc, path))
```
Comment (Suor) [inline tests/unit/remote/ssh/test_ssh.py]: Same as above. Use `remote.path_info / "some_name"`.

File: dvc/remote/ssh/__init__.py (hunk 1 above)
Comment (mroutis) [reply dvc/remote/ssh/__init__.py]: @efiop , as discussed previously, this is going to be `reflink, copy`, right?

File: dvc/remote/hdfs.py (hunk 19 above)
Comment (mroutis) [reply dvc/remote/hdfs.py]: @Suor , the assertion was moved to `repo.open`; I'll revert the commit

File: dvc/remote/local/__init__.py (hunk 18 above)
Comment (mroutis) [reply dvc/remote/local/__init__.py]: @Suor , it would be great to have a doc to put down code conventions

File: tests/func/test_data_cloud.py (hunk 25 above)
Comment (mroutis) [reply tests/func/test_data_cloud.py]: @Suor , the `remote` path is different from the one in the current directory:
```python
>>> remote.path_info.path
'/tmp/dvc-test.32712.lguntid7.CUvCFsrBf7cyEA4fndhtkW'

>>> os.getcwd()
'/tmp/dvc-test.32712.8ft94g_9.RfMwGpxMVM6wqkCHScN5bk'
```

If it shouldn't be needed, then we need to address this question:

> Why are we relying on URLInfo to make the connection and not reading from the config file or using the information of the remote?

File: dvc/remote/ssh/__init__.py [hunk 27]
```diff
@@ -164,12 +165,54 @@ def isdir(self, path_info):
...
+
+        with self.ssh(from_info) as ssh:
+            ssh.symlink(from_info.path, to_info.path)
+
+    def hardlink(self, from_info, to_info):
+        if not from_info.scheme == to_info.scheme == self.scheme:
+            raise NotImplementedError
+
+        # See dvc/remote/local/__init__.py - hardlink()
+        if self.getsize(from_info) == 0:
```
[the remark this answers is not in the document]
Comment (mroutis) [reply dvc/remote/ssh/__init__.py]: Do you mean the `self.getsize(from_info) == 0` to something like `self.is_empty(from_info)` ?

Or the whole _hardlink optimization_? I'm not sure how an abstraction will benefit here.

```python
def optimized_hardlink(from_info, to_info):
    if self.getsize(from_info) == 0:
        self.open(...)
        logger.debug(...)
       return True
    return False
```

And then call it like
```python
def hardlink(self, from_info, to_info):
     return self.optimized_hardlink(...) or System.hardlink(from_info, to_info)
```

Or maybe abstracting the `is_empty` and `create_empty`/`touch` operation.


I don't think it needs abstraction, @Suor , but I might be missing something

**Comment by efiop** [comment]
@mroutis Note that tests have failed.

**Comment by mroutis** [comment]
> @mroutis Note that tests have failed.

@efiop , yes :( still trying to find a way how to deal with `the w` and `wb` modes in `ssh.open`

**Comment by mroutis** [comment]
@efiop , I can skip the support for them and return back the assertion for read-only modes, there's a workaround (using `sftp.open` directly) that I can use instead

**Comment by efiop** [comment]
@mroutis Not sure I follow. Are you saying that
```
            self.open(to_info, "w").close()
```
causes problems because `w` is not supported, but `wb` is?

**Comment by mroutis** [comment]
@efiop , yes .

The problem is with `io.TextIOWrapper` that it doesn't support the `write_through` operation prior python 3.3 (if I don't use that one, it doesn't write the content to the file)

**Comment by efiop** [comment]
@mroutis why not use `wb` instead then?

File: dvc/remote/ssh/__init__.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/ssh/__init__.py]: We've discussed it previously, we can leave it as is and abstract it later. This is pretty good as is, and we can move forward without it for now.

File: tests/unit/remote/ssh/test_connection.py [hunk 28]
```diff
@@ -64,3 +76,28 @@ def test_walk(tmp_path, ssh):
             paths.add(posixpath.join(root, entry))
 
     assert paths == expected
+
+
+@pytest.mark.skipif(
+    platform.system() != "Darwin", reason="Only works on OSX (because of APFS)"
+)
+def test_reflink(chdir_tmp, foo, ssh):
```
Comment (efiop) [inline tests/unit/remote/ssh/test_connection.py]: you could simply use `repo_dir` instead of `chdir_tmp, foo`

Comment (mroutis) [reply tests/unit/remote/ssh/test_connection.py]: yes, but I don't need all the repository stuff :sweat_smile: I'd prefer to leave it like it is

Comment (efiop) [reply tests/unit/remote/ssh/test_connection.py]: @mroutis but you are creating new fixtures for no reason, it is better to reuse the already existing ones. What if you would need `bar`, you would create yet another fixture? Sounds bad.

Comment (efiop) [reply tests/unit/remote/ssh/test_connection.py]: Or, I guess, if we would to go that way, we would probably make repo_dir use foo fixture, but that is quite a hustle :(

Comment (mroutis) [reply tests/unit/remote/ssh/test_connection.py]: well, it's not for no reason, it saves 12 lines :sweat_smile: 

I try to stay away from `repo_dir` in unit tests, since the requirements to write a tests should be minimal; if I change `foo` for `repo_dir` I'd need to remove `repo_dir.BAR` every time (or use another name, like `repo_dir / link`)

File: tests/unit/remote/ssh/test_connection.py [hunk 29]
```diff
@@ -64,3 +72,34 @@ def test_walk(tmp_path, ssh):
...
+)
+def test_reflink(repo_dir, ssh):
+    ssh.reflink("foo", "link")
+    assert filecmp.cmp("foo", "link")
+    assert not System.is_symlink("link")
+    assert not System.is_hardlink("link")
+
+
+@pytest.mark.skipif(
+    platform.system() != "Windows",
```
Comment (efiop) [inline tests/unit/remote/ssh/test_connection.py]:
```suggestion
    platform.system() == "Windows",
```

**Comment by efiop** [comment]
@mroutis Ok, so hardlink and copy tests are also failing on 3.7 on windows. I've just realized that we are not actually testing external deps/outs scenario (see us using _should_test_ssh in TestExternalSSH in tests/func/test_repro.py). It is very interesting that they pass on 2.7, though. So there might be a can of worms that we will open. Let's skip those tests on windows too for now and create a ticket for that to take a look at later.
